repo/fsrepo/fsrepo.go

```diff
@@ -34,55 +31,74 @@ var (
 
 	// packageLock must be held to while performing any operation that modifies an
 	// FSRepo's state field. This includes Init, Open, Close, and Remove.
-	packageLock sync.Mutex // protects openersCounter and lockfiles
-	// lockfiles holds references to the Closers that ensure that repos are
-	// only accessed by one process at a time.
-	lockfiles map[string]io.Closer
-	// openersCounter prevents the fsrepo from being removed while there exist open
-	// FSRepo handles. It also ensures that the Init is atomic.
-	//
-	// packageLock also protects numOpenedRepos
-	//
-	// If an operation is used when repo is Open and the operation does not
-	// change the repo's state, the package lock does not need to be acquired.
-	openersCounter *counter.Openers
-
-	// protects dsOpenersCounter and datastores
-	dsLock           sync.Mutex
-	dsOpenersCounter *counter.Openers
-	datastores       map[string]ds2.ThreadSafeDatastoreCloser
+	packageLock sync.Mutex
 )
 
-func init() {
-	openersCounter = counter.NewOpenersCounter()
-	lockfiles = make(map[string]io.Closer)
-
-	dsOpenersCounter = counter.NewOpenersCounter()
-	datastores = make(map[string]ds2.ThreadSafeDatastoreCloser)
-}
-
 // FSRepo represents an IPFS FileSystem Repo. It is safe for use by multiple
 // callers.
 type FSRepo struct {
-	// state is the FSRepo's state (unopened, opened, closed)
-	state state
+	// has Close been called already
+	closed bool
 	// path is the file-system path
 	path string
-	// config is set on Open, guarded by packageLock
-	config *config.Config
+	// lockfile is the file system lock to prevent others from opening
+	// the same fsrepo path concurrently
+	lockfile io.Closer
+	config   *config.Config
 	// ds is set on Open
 	ds ds2.ThreadSafeDatastoreCloser
 }
 
 var _ repo.Repo = (*FSRepo)(nil)
 
-// At returns a handle to an FSRepo at the provided |path|.
-func At(repoPath string) *FSRepo {
-	// This method must not have side-effects.
-	return &FSRepo{
-		path:  path.Clean(repoPath),
-		state: unopened, // explicitly set for clarity
+// Open the FSRepo at path. Returns an error if the repo is not
+// initialized.
+func Open(repoPath string) (*FSRepo, error) {
+	packageLock.Lock()
+	defer packageLock.Unlock()
+
+	expPath, err := u.TildeExpansion(path.Clean(repoPath))
+	if err != nil {
+		return nil, err
+	}
+
+	r := &FSRepo{
+		path: expPath,
+	}
+
+	lockfile, err := lockfile.Lock(r.path)
+	if err != nil {
+		return nil, err
+	}
+	defer func() {
+		if lockfile != nil {
```

Comment:
New version of this piggybacks on `err`, and avoids the `lockfile` variable completely, how's that:

```go
       defer func() {
               // unlock on error, leave it locked on success
               if err != nil {
                       r.lockfile.Close()
               }
       }()
```